### PR TITLE
Add system to ensure conservation of variable integrals upon mesh coarsening

### LINKED ISSUE #8349: Adaptivity (coarsening) violates solute conservation
### Description of the enhancement or error report
If coarsening happens in a phase field simulation conserved order parameters do not have their value perfectly conserved, as the integral over the coarse element is likely to be different than the integral over the refined children.
I would like to attempt to fix this by adding source terms that exactly compensate the local non-conservation.

### Rationale for the enhancement or information for reproducing the error
Solute conservation is essential to produce reproducable results between runs with and without adaptivity. Solute drift affects the system's free energy considerably.

### Identified impact
Eliminate the unexpected solute drift. Existing simulations are unaffected.

### PR DESCRIPTION
**Pull request by dschwen**
Closes #8349

### CODE REVIEW

File: framework/include/base/FEProblemBase.h [hunk 1]
```diff
@@ -1064,6 +1064,9 @@ class FEProblemBase :
    */
   void reportMooseObjectDependency(MooseObject * a, MooseObject * b);
 
+  /// force the changed element lists to be cached after adaptivity steps
+  void requestCacheMeshChanged() { _cache_mesh_changed = true; }
```
Comment (friedmud) [inline framework/include/base/FEProblemBase.h]: Small thing: but I don't really like this name... it doesn't mean anything.  Does it cache that the mesh has changed?  Does it request the cache for the mesh changed?  Does it turn on caching of when the mesh changed?

Oh - it causes the elements that have been changed to be cached!  😉 

How about: `forceCachingOfChangedElementsWhenMeshChanges()`... or something...

File: framework/src/base/FEProblemBase.C [hunk 2]
```diff
@@ -160,7 +160,8 @@ FEProblemBase::FEProblemBase(const InputParameters & parameters) :
     _force_restart(getParam<bool>("force_restart")),
     _fail_next_linear_convergence_check(false),
     _currently_computing_jacobian(false),
-    _started_initial_setup(false)
+    _started_initial_setup(false),
+    _cache_mesh_changed(false)
```
Comment (friedmud) [inline framework/src/base/FEProblemBase.C]: See - this is what I mean: `cache_mesh_changed` sounds like it's going "cache whether or not the mesh changed"

**Review by friedmud — APPROVED** [review]
If this works for you... then I have no complaints.

I don't necessarily think it's the right approach though.  I think we should pursue the ability to do conservative projections during adaptivity.  This would be something for @roystgnr to help us with.

**Comment by dschwen** [comment]
Cache mesh changes

**Comment by dschwen** [comment]
Btw this only works in the simple test. In a real life parallel example
weird stuff happens. I'll work on it some more this coming week.

**Comment by dschwen** [comment]
I'll reopen when I get this fully working

**Pull request closed without merging**